core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -123,8 +124,14 @@ public TableScan useSnapshot(long scanSnapshotId) {
         "Cannot override snapshot, already set to id=%s", context.snapshotId());
     Preconditions.checkArgument(ops.current().snapshot(scanSnapshotId) != null,
         "Cannot find snapshot with ID %s", scanSnapshotId);
-    return newRefinedScan(
-        ops, table, schema, context.useSnapshotId(scanSnapshotId));
+    if (this instanceof DataTableScan) {
```

Comment:
Instead of checking for a specific subclass, I think it is cleaner to override the method in `DataTableScan`. Here's what I was able to get working:

```java
  @Override
  public TableScan useSnapshot(long scanSnapshotId) {
    TableScan newScan = super.useSnapshot(scanSnapshotId);

    if (newScan.snapshot().schemaId() != null) {
      Schema snapshotSchema = SnapshotUtil.schemaFor(table(), scanSnapshotId);
      return newRefinedScan(tableOps(), table(), snapshotSchema, context().useSnapshotId(scanSnapshotId));
    }

    return newScan;
  }
```

Reply:
Ok, I'll override the method in `DataTableScan`.

Reply:
Q. (In your example) why check if `newScan.snapshot().schemaId() != null`? We can simply call `SnapshotUtil.schemaFor(table(), scanSnapshotId)` without checking, since the util method currently returns the table schema as a fallback, pending a TODO to recover the schema for snapshot by reading old metadata. This way, once that TODO is implemented, we don't need to revisit this code.
We don't need `newScan` at all; the only reason for it would be to call `super.useSnapshot` for its side-effects (argument validation), which I prefer to avoid.

Reply:
The purpose of calling `newScan` was to use the argument validation in `BaseTableScan`. I'd rather not update that and not know that it needs to be updated here as well.